File: mapstruct/mapstruct-java-extension/src/main/java/ru/tinkoff/kora/mapstruct/java/extension/MapstructKoraExtensionFactory.java
package ru.tinkoff.kora.mapstruct.java.extension;

import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionFactory;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import java.util.Optional;

import static ru.tinkoff.kora.mapstruct.java.extension.MapstructKoraExtension.MAPPER_ANNOTATION;

public final class MapstructKoraExtensionFactory implements ExtensionFactory {
    @Override
    public Optional<KoraExtension> create(ProcessingEnvironment processingEnvironment) {
        var mapper = processingEnvironment.getElementUtils().getTypeElement(MAPPER_ANNOTATION.canonicalName());
        if (mapper == null) {
            return Optional.empty();
        }
        return Optional.of(new MapstructKoraExtension(processingEnvironment));
    }
}

File: mapstruct/mapstruct-ksp-extension/src/main/java/ru/tinkoff/kora/mapstruct/ksp/extension/MapstructKoraExtensionFactory.kt
package ru.tinkoff.kora.mapstruct.ksp.extension

import com.google.devtools.ksp.getClassDeclarationByName
import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionFactory
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension

class MapstructKoraExtensionFactory : ExtensionFactory {

    override fun create(resolver: Resolver, kspLogger: KSPLogger, codeGenerator: CodeGenerator): KoraExtension? {
        return resolver.getClassDeclarationByName(MapstructKoraExtension.mapperAnnotation.canonicalName)
            ?.let { MapstructKoraExtension }
    }
}
